File: companies/Reddit/186.reverse-words-in-a-string-ii.java
/**
 * Algorithm:
 * Reverse the entire character array, then reverse each individual word. This
 * changes word order in place while preserving each word's letters.
 *
 * Complexity:
 * Time O(n), space O(1).
 */
class Solution {
    public void reverseWords(char[] s) {
        reverse(s, 0, s.length - 1);
        int start = 0;
        for (int i = 0; i <= s.length; i++) {
            if (i == s.length || s[i] == ' ') {
                reverse(s, start, i - 1);
                start = i + 1;
            }
        }
    }

    private void reverse(char[] s, int left, int right) {
        while (left < right) {
            char tmp = s[left];
            s[left++] = s[right];
            s[right--] = tmp;
        }
    }
}

File: ServiceNow/186.reverse-words-in-a-string-ii.cpp
#include <algorithm>
#include <array>
#include <cctype>
#include <climits>
#include <cstdlib>
#include <functional>
#include <numeric>
#include <queue>
#include <sstream>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>
using namespace std;


class Solution {
public:
    void reverseWords(vector<char>& s) {
        /*
        Approach: reverse the whole character array, then reverse each individual
        word. The first reversal puts words in reverse order but with letters
        reversed; the second pass restores letters inside each word.

        C++ notes: vector<char>& mutates the input buffer in place, matching the
        Python list[str] character array.
        Complexity: O(n) time, O(1) space.
        */
        auto rev = [&](int left, int right) {
            while (left < right) swap(s[left++], s[right--]);
        };
        rev(0, (int)s.size() - 1);
        int start = 0;
        for (int i = 0; i <= (int)s.size(); ++i) {
            if (i == (int)s.size() || s[i] == ' ') {
                rev(start, i - 1);
                start = i + 1;
            }
        }
    }
};
